adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/sasl/AmqpAdapterSaslAuthenticatorFactory.java

```diff
@@ -0,0 +1,214 @@
+
+                    }
+                    completionHandler.handle(Boolean.TRUE);
+                });
+
+                final byte[] saslResponse = new byte[sasl.pending()];
+                sasl.recv(saslResponse, 0, saslResponse.length);
+
+                verifyPlain(remoteMechanism, saslResponse, deviceAuthTracker.completer());
+            }
+        }
+
+        @Override
+        public boolean succeeded() {
+            return succeeded;
+        }
+
+        private void verifyPlain(final String remoteMechanism, final byte[] saslResponse,
+                final Handler<AsyncResult<Device>> completer) {
+
+            getFields(saslResponse).compose(fields -> {
+                final DeviceCredentials credentials = UsernamePasswordCredentials
+                        .create(fields[0], fields[1], config.isSingleTenant());
+                if (credentials == null) {
+                    // adapter is configured for multi-tenancy but credentials
+                    // does not comply with the pattern [<authId>@<tenantId>]
+                    completer.handle(Future.failedFuture(new CredentialException(
+                            "username does not comply with expected pattern [<authId>@<tenantId>]")));
+                } else {
+                    authProvider.authenticate(credentials, completer);
+                }
+                return Future.future();
+            }).recover(t -> {
+                completer.handle(Future.failedFuture(t));
+                return Future.failedFuture(t);
+            });
+
+        }
+
+        private Future<String[]> getFields(final byte[] saslResponse) {
+            final Future<String[]> result = Future.future();
+            final int authzidEndPosition = findNullByte(saslResponse, 0);
+            if (authzidEndPosition < 0) {
+                result.fail(
+                        new CredentialException("PLAIN response is malformed (does not contain an Authorization ID"));
+            }
+            final int authcidEndPosition = findNullByte(saslResponse, authzidEndPosition + 1);
+            if (authcidEndPosition < 0) {
+                result.fail(new CredentialException("PLAIN response must contain an authentication ID"));
+            }
+            if (authcidEndPosition == saslResponse.length - 1) {
+                result.fail(new CredentialException("PLAIN response must contain a password"));
+            }
+            // Get the authentication identity and password (ignoring authorization id)
+            final String authcid = new String(saslResponse, authzidEndPosition + 1,
+                    authcidEndPosition - (authzidEndPosition + 1), StandardCharsets.UTF_8);
+            final String passwd = new String(saslResponse, authcidEndPosition + 1,
+                    saslResponse.length - (authcidEndPosition + 1));
+
+            result.complete(new String[] { authcid, passwd });
+
+            return result;
+        }
```

Comment:
all of this code for parsing the client's response data is a duplication of existing code (org.eclipse.hono.service.auth.AbstractHonoAuthenticationService), isn't it? Can't we find a way to put it into a place where we can re-use it properly?

Reply:
Hmm, have you done any testing regarding the performance of both variants? I am usually cautious with statements regarding the runtime performance of code because you most of the times do not know what the VM actually makes out of the code.

Anyway, I do not think that we need multiple implementations of parsing the SASL response.

Reply:
github comment boxes are a bit confusing - some of your comments are visible below with no reply-box